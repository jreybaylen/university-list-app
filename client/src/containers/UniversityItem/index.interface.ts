import { UniversityProps } from '@interface/api.interface'

interface ModifiedUniversityProps extends UniversityProps {
    onSelect?(param: string): void
}

export type {
    ModifiedUniversityProps
}