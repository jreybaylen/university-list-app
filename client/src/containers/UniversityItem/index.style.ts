import { StypesProps } from '@interface/style.interface'

const styles: StypesProps = {
    country: {
        fontSize: '.85rem',
        margin: '0 0 8px 10px'
    },
    name: {
        fontWeight: 500,
        color: '#3e3e3e',
        margin: '0 0 15px 10px'
    },
    websiteContainer: {
        display: 'flex'
    }
}

export {
    styles
}