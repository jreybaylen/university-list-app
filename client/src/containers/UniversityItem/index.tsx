import { useEffect, useState } from 'react'

import { styles } from './index.style'
import { ModifiedUniversityProps } from './index.interface'

import { Card, WebsiteLink } from '@components/index'

function University (props: ModifiedUniversityProps): JSX.Element {
    const [ name, setName ] = useState<string>('')
    const [ country, setCountry ] = useState<string>('')
    const [ websites, setWebsites ] = useState<Array<string>>([])
    const handleCardClick = () => {
        if (props.onSelect) {
            props.onSelect(props.name)
        }
    }
    const handleOpenWebsite = (website: string) => {
        window.open(website, '_blank')
    }

    useEffect(() => {
        const { name, country, ...rest } = props

        setName(name)
        setCountry(country)
        setWebsites(rest.web_pages)
    }, [ props ])
    
    const universityElement = (
        <Card onClick={ handleCardClick }>
            <p style={ styles.country }>{ country }</p>
            <h3 style={ styles.name }>{ name }</h3>
            <div style={ styles.websiteContainer }>
                { websites.map(
                    (website: string) => (
                        <WebsiteLink
                            key={ website }
                            website={ website }
                            onSelect={ handleOpenWebsite }
                        />
                    )
                ) }
            </div>
        </Card>
    )

    return universityElement
}

export default University