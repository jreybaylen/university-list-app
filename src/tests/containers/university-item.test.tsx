import { render, fireEvent } from '@testing-library/react'

import UniversityItem from '@container/UniversityItem'

const stateProvince = ''
const alpha_two_code = 'PH'
const country = 'Philippines'
const name = 'Adamson University'
const domains = [
    'https://test-url.com'
]
const web_pages = [
    'https://test-url.com'
]
const university = {
    name,
    domains,
    country,
    web_pages,
    alpha_two_code,
    'state-province': stateProvince
}

describe('<Banner /> Component', () => {
    beforeAll(() => {
        global.open = jest.fn()
    })

    it('Should render without crashing', () => {
        render(
            <UniversityItem { ...university } />
        )
    })

    it('Should render the info inside of Card component', () => {
        const { getByText } = render(
            <UniversityItem { ...university } />
        )

        expect(getByText(name)).toBeInTheDocument()
        expect(getByText(country)).toBeInTheDocument()
        expect(getByText(web_pages[0])).toBeInTheDocument()
    })

    it('Should open website link once it clicked', () => {
        const { getByText } = render(
            <UniversityItem { ...university } />
        )

        expect(global.open).not.toHaveBeenCalled()
        fireEvent.click(getByText(web_pages[0]))
        expect(global.open).toHaveBeenCalled()
    })

    it('Should trigger onSelect once the body of the card has been clicked', () => {
        const handleSelectCard = jest.fn()
        const { getByText } = render(
            <UniversityItem { ...university } onSelect={ handleSelectCard } />
        )

        expect(handleSelectCard).not.toHaveBeenCalled()
        fireEvent.click(getByText(country))
        expect(handleSelectCard).toHaveBeenCalled()
    })
})