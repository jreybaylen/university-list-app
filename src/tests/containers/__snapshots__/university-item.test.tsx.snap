// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Banner /> Component Should render without crashing 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        style="margin: 15px; padding: 30px; border-radius: 5px; cursor: pointer; box-shadow: none;"
      >
        <p
          style="font-size: .85rem; margin: 0px 0px 8px 10px;"
        >
          Philippines
        </p>
        <h3
          style="font-weight: 500; color: rgb(62, 62, 62); margin: 0px 0px 15px 10px;"
        >
          Adamson University
        </h3>
        <div
          style="display: flex;"
        >
          <button
            style="cursor: pointer; margin-right: 10px; border-radius: 4px; color: rgb(33, 150, 243); font-size: .95rem; padding: 6px 10px; text-decoration: none;"
          >
            https://test-url.com
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      style="margin: 15px; padding: 30px; border-radius: 5px; cursor: pointer; box-shadow: none;"
    >
      <p
        style="font-size: .85rem; margin: 0px 0px 8px 10px;"
      >
        Philippines
      </p>
      <h3
        style="font-weight: 500; color: rgb(62, 62, 62); margin: 0px 0px 15px 10px;"
      >
        Adamson University
      </h3>
      <div
        style="display: flex;"
      >
        <button
          style="cursor: pointer; margin-right: 10px; border-radius: 4px; color: rgb(33, 150, 243); font-size: .95rem; padding: 6px 10px; text-decoration: none;"
        >
          https://test-url.com
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
